flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java

```diff
@@ -82,4 +91,78 @@ public static ResolveOrder fromString(String resolveOrder) {
 			super(urls, parent);
 		}
 	}
+
+	/**
+	 * Ensures that holding a reference on the context class loader outliving the scope of user code does not prevent
+	 * the user classloader to be garbage collected (FLINK-16245).
+	 *
+	 * <p>This classloader delegates to the actual user classloader. Upon {@link #close()}, the delegate is nulled
+	 * and can be garbage collected. Additional class resolution will be resolved solely through the bootstrap
+	 * classloader and most likely result in ClassNotFound exceptions.
+	 *
+	 * @param <T> the classloader type that also needs to be closeable.
+	 */
+	private static class SafetyNetWrapperClassLoader<T extends ClassLoader & Closeable> extends CloseableClassLoader
+			implements Closeable {
+		private static final Logger LOG = LoggerFactory.getLogger(SafetyNetWrapperClassLoader.class);
+
+		private T inner;
+
+		SafetyNetWrapperClassLoader(T inner) {
+			super(null);
+			this.inner = inner;
+		}
+
+		@Override
+		public void close() {
+			if (inner != null) {
+				try {
+					inner.close();
+				} catch (IOException e) {
+					LOG.warn("Could not close user classloader", e);
+				}
+			}
+			inner = null;
+		}
+
+		@Override
+		protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
+			if (inner == null) {
+				return super.loadClass(name, resolve);
+			}
+
+			synchronized (getClassLoadingLock(name)) {
+				final Class<?> loadedClass = findLoadedClass(name);
+				if (loadedClass != null) {
+					return resolveIfNeeded(resolve, loadedClass);
+				}
+
+				return resolveIfNeeded(resolve, inner.loadClass(name));
+			}
+		}
+
+		private Class<?> resolveIfNeeded(final boolean resolve, final Class<?> loadedClass) {
```

Comment:
why is this method necessary? Shouldn't the inner classloader handle this?

Reply:
`loadClass(String name, boolean resolve)` is protected, `loadClass(String name)` is not resolving by default.

Reply:
Now that the safety net is a URLClassLoader we can just forward the call, correct?

Reply:
No, still not possible.

Reply:
Added a common ancestor now.